parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java

```diff
@@ -237,44 +254,108 @@ private void writeUnion(GroupType parquetSchema, Schema avroSchema,
     recordConsumer.endGroup();
   }
 
-  @SuppressWarnings("unchecked")
+  /**
+   * Calls an appropriate write method based on the value.
+   * Value MUST not be null.
+   *
+   * @param type the Parquet type
+   * @param avroSchema the Avro schema
+   * @param value a non-null value to write
+   */
   private void writeValue(Type type, Schema avroSchema, Object value) {
     Schema nonNullAvroSchema = AvroSchemaConverter.getNonNull(avroSchema);
-    Schema.Type avroType = nonNullAvroSchema.getType();
-    if (avroType.equals(Schema.Type.BOOLEAN)) {
-      recordConsumer.addBoolean((Boolean) value);
-    } else if (avroType.equals(Schema.Type.INT)) {
-      if (value instanceof Character) {
-        recordConsumer.addInteger((Character) value);
-      } else {
-        recordConsumer.addInteger(((Number) value).intValue());
-      }
-    } else if (avroType.equals(Schema.Type.LONG)) {
-      recordConsumer.addLong(((Number) value).longValue());
-    } else if (avroType.equals(Schema.Type.FLOAT)) {
-      recordConsumer.addFloat(((Number) value).floatValue());
-    } else if (avroType.equals(Schema.Type.DOUBLE)) {
-      recordConsumer.addDouble(((Number) value).doubleValue());
-    } else if (avroType.equals(Schema.Type.BYTES)) {
-      if (value instanceof byte[]) {
-        recordConsumer.addBinary(Binary.fromReusedByteArray((byte[]) value));
-      } else {
-        recordConsumer.addBinary(Binary.fromReusedByteBuffer((ByteBuffer) value));
-      }
-    } else if (avroType.equals(Schema.Type.STRING)) {
-      recordConsumer.addBinary(fromAvroString(value));
-    } else if (avroType.equals(Schema.Type.RECORD)) {
-      writeRecord(type.asGroupType(), nonNullAvroSchema, value);
-    } else if (avroType.equals(Schema.Type.ENUM)) {
-      recordConsumer.addBinary(Binary.fromString(value.toString()));
-    } else if (avroType.equals(Schema.Type.ARRAY)) {
-      listWriter.writeList(type.asGroupType(), nonNullAvroSchema, value);
-    } else if (avroType.equals(Schema.Type.MAP)) {
-      writeMap(type.asGroupType(), nonNullAvroSchema, (Map<CharSequence, ?>) value);
-    } else if (avroType.equals(Schema.Type.UNION)) {
-      writeUnion(type.asGroupType(), nonNullAvroSchema, value);
-    } else if (avroType.equals(Schema.Type.FIXED)) {
-      recordConsumer.addBinary(Binary.fromReusedByteArray(((GenericFixed) value).bytes()));
+    LogicalType logicalType = nonNullAvroSchema.getLogicalType();
```

Comment:
No, we do the same thing in Avro. That's why we had to store the logical type as a field on the schema (rather than caching after parsing), but it didn't slow things down.